addons/web_editor/static/src/js/editor/snippets.options.js

```diff
@@ -2513,6 +2519,395 @@ registry['sizing_y'] = registry.sizing.extend({
     },
 });
 
+/*
+ * Abstract option to be extended by the ImageOptimize and BackgroundOptimize
+ * options that handles all the common parts.
+ */
+const ImageHandlerOption = SnippetOptionWidget.extend({
+    events: {
+        'change [data-name=image_quality_opt] input': '_onQualityChange',
+    },
+
+    /**
+     * @override
+     */
+    async willStart() {
+        const _super = this._super.bind(this);
+        await this._loadImageInfo();
+        // Make sure image is loaded because we need its naturalWidth to render XML
+        const img = this._getImg();
+        await new Promise((resolve, reject) => {
+            if (img.complete) {
+                return resolve();
+            }
+            img.addEventListener('load', resolve, {once: true});
+            img.addEventListener('error', resolve, {once: true});
+        });
+        return _super(...arguments);
+    },
+
+    //--------------------------------------------------------------------------
+    // Options
+    //--------------------------------------------------------------------------
+
+    /**
+     * @see this.selectClass for parameters
+     */
+    selectWidth(previewMode, widgetValue, params) {
+        this._getImg().dataset.resizeWidth = widgetValue;
+        return this._applyOptions();
+    },
+    /**
+     * @see this.selectClass for parameters
+     */
+    setFilter(previewMode, widgetValue, params) {
+        this._getImg().dataset.filter = this._normalizeColor(widgetValue);
+        return this._applyOptions();
+    },
+
+    //--------------------------------------------------------------------------
+    // Private
+    //--------------------------------------------------------------------------
+
+    /**
+     * @override
+     */
+    _computeVisibility() {
+        const src = this._getImg().getAttribute('src');
+        return src && src !== '/';
+    },
+    /**
+     * @override
+     */
+    _computeWidgetState(methodName, params) {
+        const img = this._getImg();
+        switch (methodName) {
+            case 'selectWidth':
+                return img.naturalWidth;
+            case 'setFilter':
+                return img.dataset.filter;
+        }
+        return this._super(...arguments);
+    },
+    /**
+     * @override
+     */
+    async _renderCustomXML(uiFragment) {
+        if (!this.originalSrc) {
+            return [...uiFragment.childNodes].forEach(node => {
```

Comment:
Returing a `forEach` result... I think you can use the `for (const ... of ...) {` form to avoid the array building and then `return;` after it ?

Reply:
The array building is because childNodes is a live collection, removing a node will remove it from the collection but the loop counter will continue, so you will miss every other element if you iterate directly over it, which is why I create a copy of the childNodes in an array. I can move the return after the forEach if you would like though.

Reply:
Indeed I missed that point. For the return, I guess this is personal taste so do as you wish.

Reply:
@qsm-odoo  Quel point ? 
Il y en a 4 sur la ligne:
 `return [...uiFragment.childNodes].forEach`

Ok, j'arrete poru aujourd'hui ;)